### ISSUE TITLE
Unable to resolve a host name to an IP address

### ISSUE THREAD

**Issue by tcleenewerck** [pharo-project/pharo-vm#220]
I cannot connect to a firebase cloud database. The following code

```
(ZnEasy get: 'https://digitalelectricmeter-default-rtdb.europe-west1.firebasedatabase.app/year/2021/2021-04-22.json') entity. 
```

should return the following response

```
"a ZnStringEntity(application/json;charset=utf-8 36B {
  ""error"" : ""Permission denied""
}
)"
```

Instead, I get an error on the UI screen: 'NameLookupFailure: cannot resolve '...' .

When I try to resolve a host name using the NetNameResolver I get the same error.

```
NetNameResolver addressForName: 'digitalelectricmeter-default-rtdb.europe-west1.firebasedatabase.app' timeout: 30. 
```

I'm using a Pharo 8.0-32bit image and Pharo8.0-32bit-7dc39ed.sources . The image is running on the arm-vm. 

When I use chromium on the pi, or curl on the pi, or ping on the pi, or host on the pi, then the host or url works. 
Other urls on the pi using the Zink or NetNameResolver in the image work correctly.

I performed the following test on a blank installation (Pharo 8.0 - 32 bit image - vm-arm):

1/ `wget -O - pharoiot.org/install | bash /dev/stdin https://github.com/pharo-iot/PharoThings/releases/download/v0.3.2/Pharo8.0-32bit-PharoThings.zip`
2/ `./pharo-ui`
3/ open a playground and run the NetNameResolver code.

I performed the following test on a blank installation (Pharo7, 32 bit image, ARM VM) following the instructions on http://get.pharoiot.org/ .

1/ wget -O - get.pharoiot.org/server | bash
2/ `./pharo-ui`
3/ open a playground and run the NetNameResolver code.

**Comment by tcleenewerck** [comment]
O yes. I included the url of my database so that you guys can test it. I would appreciate that you only use the url to investigate the problem.

**Comment by astares** [comment]
Windows 10 with latest VM + latest Pharo 9 image, no additional proxy or other:

![image](https://user-images.githubusercontent.com/5980033/116380789-9787be00-a814-11eb-8cf7-ee523458d2e5.png)

**Comment by tcleenewerck** [comment]
@astares Indeed. I'm not surprised. On my desktop on pharo 8 it works as well ;-)

**Comment by astares** [comment]
So it seems like its an **arm-vm** related problem

**Comment by tesonep** [comment]
It looks like a problem with IPv6. Is it possible that the machine has IPv6 address?

**Comment by tcleenewerck** [comment]
The machine (the raspberry pi) has a IPv4 and IPv6 address

This is the output of `ifconfig`: 

```
eth0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether b8:27:eb:b2:02:6f  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 28  bytes 1092 (1.0 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 28  bytes 1092 (1.0 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlan0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.0.222  netmask 255.255.255.0  broadcast 192.168.0.255
        inet6 2a02:1811:2c36:f100:831c:59cb:6208:6bb7  prefixlen 64  scopeid 0x0<global>
        inet6 fe80::494:9259:27a3:40d8  prefixlen 64  scopeid 0x20<link>
        ether b8:27:eb:e7:57:3a  txqueuelen 1000  (Ethernet)
        RX packets 214978  bytes 278218382 (265.3 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 82806  bytes 8419803 (8.0 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

**Comment by ezeaguerre** [comment]
Hi! Today I bumped into this issue when trying to connect to an AWS RDS database. It happens only in Linux and I think I know why.

The SocketPlugin uses the MAXHOSTNAMELEN constant as a way to limit the size of the hostname and prepare buffers big enough for the job. The problem is that on linux, this is defined as 64 (which is correct, because a hots name can't be greater than this, the problem is that we're trying to resolve a fully qualified domain name, not just the host part):

In /usr/include/sys/param.h you'll see:

```c
#if !defined MAXHOSTNAMELEN && defined HOST_NAME_MAX
# define MAXHOSTNAMELEN HOST_NAME_MAX
#endif
```

and then on /usr/include/bits/local_lim.h there's this definition:

```c
/* Maximum host name length.  */
#define HOST_NAME_MAX           64
```

However, on FreeBSD (and so, also in Mac), it's defined as 256: https://cgit.freebsd.org/src/tree/sys/sys/param.h#n134

I don't know about windows. But the socket plugins defines MAXHOSTNAMELEN as 256 if it's not defined.

You can confirm this by compiling this small example on any Linux:

```c
#include <stdio.h>
#include <sys/param.h>

int main() {
        printf( "MAXHOSTNAMELEN: %d\n", MAXHOSTNAMELEN );
        return 0;
}
```

And you can also confirm this by inspecting the binary of the socket plugin, as there's a global buffer of the size of this constant defined here: https://github.com/pharo-project/opensmalltalk-vm/blob/pharo-9/extracted/plugins/SocketPlugin/src/common/SocketPluginImpl.c#L193

```
$ readelf -s libSocketPlugin.so| grep localHostName
    89: 00000000002107e0    64 OBJECT  LOCAL  DEFAULT   24 localHostName
```

It says the size of the symbol is "64".

This same issue happened on Nmap: https://github.com/nmap/nmap/issues/140 and they fixed it by defining their own constant (on the linked PR). So, I guess we should do something like that. The most easy fix would be to undefine MAXHOSTNAMELEN and redefine it as 256 (255 for the FQDN and 1 extra for the ending 0 character).

As a temporal fix, I added an entry to my /etc/hosts with a short name, pointing to the IP of the RDS database, but I'd rather use the DNS name, so the IP can change freely (for example, during a failover switch).

Should we also report this on the upstream opensmalltalk-vm ?

**Comment by tesonep** [comment]
Fixed by PR #386
